## Task 1: Kernelized SVM



Importing the Data

In [1]:
from sklearn import datasets
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline

digits = datasets.load_digits()
n_samples = len(digits.images)
X = digits.images.reshape((n_samples, -1))
y = digits.target == 8
print(
    f"The number of images is {X.shape[0]} and each image contains {X.shape[1]} pixels"
)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

The number of images is 1797 and each image contains 64 pixels


### RBF-Kernel
<br>

Calculation of a reference accuracy without hyperparameter tuning:

In [2]:
reference_model = make_pipeline( SVC(kernel='rbf'))
reference_model.fit(X, y)
reference_model.fit(X_train, y_train)
y_pred_ref = reference_model.predict(X_test)      

print("Reference accuracy without hyperparameter tuning: "+ str( accuracy_score(y_test, y_pred_ref)) )


Reference accuracy without hyperparameter tuning: 0.987037037037037


Every fitting process is quite equivalent. At first, the Parameter gid gets defined. Then a grid search gets performed where for each parameter set and SVC gets calculated. The parameter set of the model with the best accuracy ist selected and displayed.

In [3]:
tuned_parameters = [
    {"kernel": ["rbf"], "gamma": [1e-2, 1e-4], "C": [ 1, 10, 100, 1000]}
]

grid_search = GridSearchCV(     
    SVC(), tuned_parameters)
grid_search.fit(X_train, y_train)
y_pred_test = grid_search.predict(X_test)      
y_pred_train = grid_search.predict(X_train)      
print("The best kernel has these params: " + str(grid_search.best_params_)+ "\n The model provides an test accuracy of: "+ str( accuracy_score(y_test, y_pred_test)) + " on the test set, \n as well as an train accuracy of: "+ str( accuracy_score(y_train, y_pred_train)) + " on the training set")



The best kernel has these params: {'C': 100, 'gamma': 0.0001, 'kernel': 'rbf'}
 The model provides an test accuracy of: 0.9925925925925926 on the test set, 
 as well as an train accuracy of: 0.9992044550517104 on the training set


The model with tuned parameters provides an accuracy of  0.9926 and the reference one of  0.9870. For that reason, the tuned parameter achieved a better result than the reference parameters. <br>
The parameters "C" and "gamma" were tuned with GridSearchCV. These Parameters were selected because they are the only ones influencing the RBF kernel. Other parameters like "degree" could be set but would not be considered in an RBF kernel. 

### Linear-Kernel
<br>

Calculation of a reference accuracy without hyperparameter tuning:

In [4]:
reference_model = make_pipeline( SVC(kernel='linear'))
reference_model.fit(X, y)
reference_model.fit(X_train, y_train)
y_pred_ref = reference_model.predict(X_test)      

print("Reference accuracy without hyperparameter tuning: "+ str( accuracy_score(y_test, y_pred_ref)) )

Reference accuracy without hyperparameter tuning: 0.9444444444444444


In [5]:
tuned_parameters = [
    {"kernel": ["linear"], "C": [0.001, 0.01, 0.1, 1, 10]},
]

grid_search = GridSearchCV(                                            
    SVC(), tuned_parameters     
)
grid_search.fit(X_train, y_train)
y_pred_test = grid_search.predict(X_test)      
y_pred_train = grid_search.predict(X_train)      
print("The best kernel has these params: " + str(grid_search.best_params_)+ "\n The model provides an test accuracy of: "+ str( accuracy_score(y_test, y_pred_test)) + " on the test set, \n as well as an train accuracy of: "+ str( accuracy_score(y_train, y_pred_train)) + " on the training set")


The best kernel has these params: {'C': 0.001, 'kernel': 'linear'}
 The model provides an test accuracy of: 0.9555555555555556 on the test set, 
 as well as an train accuracy of: 0.9705648369132857 on the training set


The model with tuned parameters provides an accuracy of  0.9519 and the reference one of  0.9444. For that reason, the tuned parameter achieved a better result than the reference parameters. <br>
Only the parameter was "C" tuned with GridSearchCV because the other parameters don't affect a linear Kernel.

### Poly-Kernel
<br>

Calculation of a reference accuracy without hyperparameter tuning:

In [6]:
reference_model = make_pipeline( SVC(kernel='poly'))
reference_model.fit(X, y)
reference_model.fit(X_train, y_train)
y_pred_ref = reference_model.predict(X_test)      

print("Reference accuracy without hyperparameter tuning: "+ str( accuracy_score(y_test, y_pred_ref)) )

Reference accuracy without hyperparameter tuning: 0.9907407407407407


In [7]:
tuned_parameters = [
    {"kernel": ["poly"],"coef0":[-1,0,1,10,100], "gamma": [1e-3, 1e-4], "C": [10, 100, 1000], "degree": [1, 2, 3,4,5]},
]

grid_search = GridSearchCV(                                           
    SVC(), tuned_parameters                                                                          
)
grid_search.fit(X_train, y_train)
y_pred_test = grid_search.predict(X_test)      
y_pred_train = grid_search.predict(X_train)      
print("The best kernel has these params: " + str(grid_search.best_params_)+ "\n The model provides an test accuracy of: "+ str( accuracy_score(y_test, y_pred_test)) + " on the test set, \n as well as an train accuracy of: "+ str( accuracy_score(y_train, y_pred_train)) + " on the training set")


The best kernel has these params: {'C': 10, 'coef0': 0, 'degree': 4, 'gamma': 0.001, 'kernel': 'poly'}
 The model provides an test accuracy of: 0.9944444444444445 on the test set, 
 as well as an train accuracy of: 1.0 on the training set


The model with tuned parameters provides an accuracy of  0.9944 and the reference one of  0.9907. For that reason, the tuned parameter achieved a better result than the reference parameters. <br>
The parameters "C", "gamma" and "coef0" where tuned with GridSearchCV. These Parameters were selected because they are the only ones influencing the poly kernel.

### Sigmoid-Kernel
<br>

Calculation of a reference accuracy without hyperparameter tuning:

In [8]:
reference_model = make_pipeline( SVC(kernel='sigmoid'))
reference_model.fit(X, y)
reference_model.fit(X_train, y_train)
y_pred_ref = reference_model.predict(X_test)      

print("Reference accuracy without hyperparameter tuning: "+ str( accuracy_score(y_test, y_pred_ref)) )

Reference accuracy without hyperparameter tuning: 0.8796296296296297


In [9]:
tuned_parameters = [
    {"kernel": ["sigmoid"],"coef0":[-10,-1,0,1,10,100], "gamma": [1e-3, 1e-4], "C": [10, 100, 1000]},
]

grid_search = GridSearchCV(                                        
    SVC(), tuned_parameters                                                                
)
grid_search.fit(X_train, y_train)
y_pred_test = grid_search.predict(X_test)      
y_pred_train = grid_search.predict(X_train)      
print("The best kernel has these params: " + str(grid_search.best_params_)+ "\n The model provides an test accuracy of: "+ str( accuracy_score(y_test, y_pred_test)) + " on the test set, \n as well as an train accuracy of: "+ str( accuracy_score(y_train, y_pred_train)) + " on the training set")

The best kernel has these params: {'C': 1000, 'coef0': -1, 'gamma': 0.0001, 'kernel': 'sigmoid'}
 The model provides an test accuracy of: 0.9925925925925926 on the test set, 
 as well as an train accuracy of: 1.0 on the training set


The model with tuned parameters provides an accuracy of  0.9926 and the reference one of  0.8796. For that reason, the tuned parameter achieved a better result than the reference parameters. <br>
The parameters "C", "gamma" and "coef0" where tuned with GridSearchCV. These Parameters were selected because they are the only ones influencing the poly kernel.